### ISSUE TITLE
Bug in `mul_rev`

### ISSUE THREAD

**Issue by dpsanders** [JuliaIntervals/IntervalContractors.jl#15]
```jl
julia> c = -100..100
[-100, 100]

julia> mul_rev(0..0, 2..2, c)
([0, 0], ∅, ∅)
```
This should return `0` for the third argument.
The problem is that `b` is modified within the function and becomes empty, and then that empty value is used to calculate `c`.

**Comment by mewilhel** [comment]
Just took a swing at this. I think the issue may also be related to the fact that `b = b ∩ (a / c)` requires the evaluate of `a/c`  or in this example `0..0/-100..100` which contains the form `0/0`.  So I've included a version that catches this case and doesn't refine the argument that would give rise to a calculation contains `0/0`. I think we'd need to define division in terms of an operator that creates a union of two disjoint intervals otherwise. What are your thoughts on the below?

```julia
function mul_rev1(a::Interval, b::Interval, c::Interval)  # a = b * c
    temp = deepcopy(b)
    b = b ∩ (a / c)
    c = c ∩ (a / temp)
    return a, b, c
end

function mul_rev2(a::Interval, b::Interval, c::Interval)
    if (in(0.0,c) && ~in(0.0,b))
        c = c ∩ (a / b)
    elseif (~in(0.0,c) && in(0.0,b))
        b = b ∩ (a / c)
    elseif (~in(0.0,c) && ~in(0.0,b))
        b = b ∩ (a / c)
        c = c ∩ (a / b)
    end
return a, b, c
end
```
Testing these two variants out gives:
```julia
mul_rev1(0..0, 2..2, c)
([0, 0], ∅, [0, 0])

mul_rev2(0..0, 2..2, c)
([0, 0], [2, 2], [0, 0])

mul_rev1(1..1, 2..2, c)
([1, 1], [2, 2], [0.5, 0.5])

mul_rev2(1..1, 2..2, c)
([1, 1], [2, 2], [0.5, 0.5])
```

**Comment by dpsanders** [comment]
Thanks for looking at this. In the master branch of IntervalArithmetic, `(0..0) / (-100..100)` gives `0..0`.

**Comment by dpsanders** [comment]
(Just a heads-up that I will tag a new version of IntervalArithmetic.jl shortly.)

**Comment by mewilhel** [comment]
Excellent to hear it! It's been a really useful package for a lot of the McCormick relaxation code I've been writing. 

I'm actually working on the master branch as well  (I defined some reverse interval operators in EaGO.jl that I'm planning on transferring to this package, hyperbolic functions, mainly. So I figured I'd work on the most current version). 

I think the '(0..0) / (-100..100) = 0..0' is actually creating this error. So in the example we have

`
b = b ∩ (a / c)
b = 2..2 ∩ ((0..0) / (-100..100))
b = 2..2 ∩ 0..0
b = ∅
`

However, if we had '(0..0) / (-100..100) = -∞..∞', we get a correct answer here. I'm not sure how changing this would play with the other ValidatedNumerics.jl functions so `mul_rev2` should be an okay work around if keeping line 174 in IntervalArithmetic.jl/blob/master/src/intervals/arithmetic.jl is desirable.

**Comment by dpsanders** [comment]
But I guess you could say that 2..2 is the correct answer, since there's *a* solution of `2c==0` in the interval `c`.

**Comment by dpsanders** [comment]
Note that you can use `∈` (type as `\in<TAB>`).

**Comment by dpsanders** [comment]
And `∉` (`\notin<TAB>`)

**Comment by dpsanders** [comment]
(Just a heads-up that there's an `extended_div` function in the master branch of IntervalArithmetic.)

**Comment by mewilhel** [comment]
Thanks for the syntax and extended_div heads up! 

I think the 2..2 interval is the correct answer. The resulting intervals are still subsets of the original A, B, and C (just not necessarily strict). After applying the reverse contractor, the resulting interval box should contain the set \{ (a,b,c) ∈ (A, B, C) such that a = b × c \} and since  2..2 × 0..0 = 0..0 this set of solutions is nonempty.

**Comment by dpsanders** [comment]
Thanks, your argument that the result should contain `{ (a,b,c) ∈ (A, B, C) such that a = b × c }` is convincing.

There was a discussion about `a / b` where `a == 0` and `b` contains zero. 
I think the conclusion was that the current behaviour (in current master) is the one mandated by the standard. cc @eeshan9815

**Comment by dpsanders** [comment]
I think the logic in `mul_rev_2` can be reduced to "do the division if the thing you're dividing by is non-zero". If we reduce to using functions that look at only a single variable, we will only need to return that variable, rather than the three variables.

**Comment by mewilhel** [comment]
Actually, I was just skimming the below article and this is addressed (pg 15-16). Looks like the a/b definition is fine for forward evaluation and they purpose two schemes for dealing with the backward propagation. The tighter of the two is the `mul_rev3` function below:

```julia
function mul_rev3(a::Interval, b::Interval, c::Interval)
     ((0.0 ∉ a) ||  (0.0 ∉ b)) && (c = c ∩ (a / b))
     ((0.0 ∉ a) ||  (0.0 ∉ c)) && (b = b ∩ (a / c))
     return a, b, c
end
```

"Vu, X. H., Schichl, H., & Sam-Haroud, D. (2009). Interval propagation and search on directed acyclic graphs for numerical constraint solving. Journal of Global Optimization, 45(4), 499."

**Comment by dpsanders** [comment]
That's useful, thanks! I added a link to a preprint version.

**Comment by eeshan9815** [comment]
I agree that the handling of contraction for each variable separately and careful usage of the API should be the preferred way.

**Comment by dpsanders** [comment]
Fixed in #18.
